lib/aemdownload.js

```diff
@@ -0,0 +1,111 @@
+        const sourceUrl = new URL(downloadFile.fileUrl);
+        const targetUrl = fileUrl(downloadFile.filePath);
+
+        const source = new Asset(sourceUrl, options.headers);
+        const target = new Asset(targetUrl);
+
+        const transferAsset = new TransferAsset(source, target, {
+            acceptRanges: true,
+            metadata: new AssetMetadata(downloadFile.filePath, undefined, downloadFile.fileSize)
+        });
+
+        yield transferAsset;
+    }
+}
+
+class AEMDownload extends EventEmitter {
+
+    /**
+     * @typedef {Object} DownloadFile
+     * @property {String} fileUrl AEM url of file to download
+     * @property {String} filePath Path on the local disk where to download
+     * @property {Number} fileSize Size of the file being downloaded
+     */
+    /**
+     * @typedef {Object} AEMDownloadOptions
+     * @property {DownloadFile[]} downloadFiles List of files that will be downloaded.
+     * @property {*} headers HTTP headers that will be included in each request sent to AEM.
+     * @property {Boolean} concurrent If true, multiple files in the supplied list of download files will transfer simultaneously. If false, only one file will transfer at a time, and the next file will not begin transferring until the current file finishes.
+     * @property {Number} maxConcurrent Maximum number of concurrent HTTP requests that are allowed
+     * @property {Number} [preferredPartSize] Preferred part size
+     */
+    /**
+     * Download files from AEM to local disk
+     * 
+     * @param {AEMDownloadOptions} options AEM download options
+     */
+    async downloadFiles(options) {
+        const preferredPartSize = options && options.preferredPartSize;
+        const maxConcurrent = (options && options.concurrent && options.maxConcurrent) || 1;
+
+        const controller = new TransferController();
+        controller.on(TransferEvents.TRANSFER_START, ({ transferAsset }) => {
+            this.emit("filestart", transferAsset.eventData);
+        });
+        controller.on(TransferEvents.TRANSFER_PROGRESS, ({ transferAsset, transferBytes }) => {
+            this.emit("fileprogress", {
+                ...transferAsset.eventData,
+                transferred: transferBytes
+            });
+        });
+        controller.on(TransferEvents.TRANSFER_COMPLETE, ({ transferAsset }) => {
+            this.emit("fileend", transferAsset.eventData);
+        });
+
+        // Build and execute pipeline
+        const randomFileAccess = new RandomFileAccess();
+        try {
+            const pipeline = new Pipeline(
+                new CreateTransferParts({ preferredPartSize }),
+                new MapConcurrent(new Transfer(randomFileAccess), { maxConcurrent }),
+                new JoinTransferParts(controller),
+                new CloseFiles(randomFileAccess),
+            );
+            await executePipeline(pipeline, generateAEMDownloadTransferRecords(options));
```

Comment:
Why not at least log the possible error in a catch statement?

Reply:
Personally, I would leave it up to the consumer to handle the error as required by their specific use-case.

Reply:
catch and rethrow, then?

Reply:
The errors are supposed to be handled as part of the transfer, and collected. This isn't fully there yet. This finally is just a final place to make sure we don't leak any file handles. If it works properly even when transfer errors occur, there won't be an exception here.